### PR TITLE
Added case in derive_by_array for sparse array

### PR DESCRIPTION
**Pull request by kangzhiq**
#### References to other Issues or PRs

#16941

#### Brief description of what is fixed or changed
The sparse array is cast to dense array while performing a derivation. Using directly sparse array will bring an improvement when it comes to a great dimension array whose value are mostly zeros. 
The sparse was cast to dense array for derivative by both **array** and **scalar**. 
Before:
```python
>>> a = MutableSparseNDimArray.zeros(10000, 20000)
>>> a[1, 1] = i
>>> a[1, 2] = j
>>> d = derive_by_array(a, i)
MemoryError
>>> d = derive_by_array(a, (i, j))
MemoryError
```
Now：
```python
>>> a = MutableSparseNDimArray.zeros(10000, 20000)
>>> a[1, 1] = i
>>> derive_by_array(a, i) = ImmutableSparseNDimArray({20001:1},(10000, 20000))
True
>>> a[1, 2] = j
>>> derive_by_array(a, (i, j)) == ImmutableSparseNDimArray({20001: 1, 200020002: 1},(2, 10000, 20000))
True
```

#### Release Notes

* tensor
  * Added specific case for sparse array in derive_by_array

### CODE REVIEW

File: sympy/tensor/array/ndim_array.py [hunk 1]
```diff
@@ -129,10 +129,18 @@ def f(pointer):
     @classmethod
     def _handle_ndarray_creation_inputs(cls, iterable=None, shape=None, **kwargs):
         from sympy.matrices.matrices import MatrixBase
+        from sympy.tensor.array import SparseNDimArray
+        from sympy import Dict
 
         if shape is None and iterable is None:
             shape = ()
             iterable = ()
+        # Construction of a sparse array from a disctionary
```
Comment (czgdp1807) [inline sympy/tensor/array/ndim_array.py]: disctionary -> dictionary

File: sympy/tensor/array/ndim_array.py [hunk 2]
```diff
@@ -129,10 +129,18 @@ def f(pointer):
...
+        elif shape is not None and isinstance(iterable, (dict, Dict)):
```
Comment (czgdp1807) [inline sympy/tensor/array/ndim_array.py]: elif -> if as there is a return under it

File: sympy/tensor/array/ndim_array.py [hunk 3]
```diff
@@ -129,10 +129,18 @@ def f(pointer):
...
+            return shape, iterable
+        # Construction of a sparse array from a sparse array
+        elif shape is None and isinstance(iterable, SparseNDimArray):
```
Comment (czgdp1807) [inline sympy/tensor/array/ndim_array.py]: > elif -> if as there is a return under it

+1

File: sympy/tensor/array/sparse_ndim_array.py [hunk 4]
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
...
+            for i in range(n):
+                s_a2 = {}
+                for key in s_a:
+                    value = _sympify(s_a[key])
+                    s_a2[key] = Derivative(value, s, **{'evaluate':True})
+                if s_a == s_a2 or s_a2 is None:
+                    break
+                s_a = s_a2
+            return type(self)(s_a, self._shape)
+        else:
```
Comment (czgdp1807) [inline sympy/tensor/array/sparse_ndim_array.py]: else not needed.

File: sympy/tensor/array/sparse_ndim_array.py [hunk 5]
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
...
+            return None
```
Comment (czgdp1807) [inline sympy/tensor/array/sparse_ndim_array.py]: Infact, AFAIK, any function/method in python returns `None` by default, so I think it's possible to remove, L127 and L128.

File: sympy/tensor/array/arrayop.py [hunk 6]
```diff
@@ -197,7 +198,8 @@ def derive_by_array(expr, dx):
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: have you tried `.as_immutable()`?

File: sympy/tensor/array/ndim_array.py [hunk 7]
```diff
@@ -129,10 +129,18 @@ def f(pointer):
     @classmethod
     def _handle_ndarray_creation_inputs(cls, iterable=None, shape=None, **kwargs):
         from sympy.matrices.matrices import MatrixBase
+        from sympy.tensor.array import SparseNDimArray
+        from sympy import Dict
 
         if shape is None and iterable is None:
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: maybe group `shape is None` into an external `if`

File: sympy/tensor/array/sparse_ndim_array.py [hunk 8]
```diff
@@ -109,7 +109,24 @@ def reshape(self, *newshape):
         if new_total_size != self._loop_size:
             raise ValueError("Invalid reshape parameters " + newshape)
 
-        return type(self)(*(newshape + (self._array,)))
+        return type(self)(*(newshape + (self._sparse_array,)))
+    
+    def _eval_derivative_n_times(self, s, n):
```
Comment (Upabjojr) [inline sympy/tensor/array/sparse_ndim_array.py]: maybe you should not overwrite this specific method. Try to look for `_visit_eval_derivative_<...>` methods.

File: sympy/tensor/array/ndim_array.py (hunk 2 above)
Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: Is it a convention in SymPy?

Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: Is this a convention in SymPy?

File: sympy/tensor/array/sparse_ndim_array.py (hunk 8 above)
Comment (kangzhiq) [reply sympy/tensor/array/sparse_ndim_array.py]: I believe that is it `al_derivative_array` that should be overridden.

File: sympy/tensor/array/ndim_array.py (hunk 2 above)
Comment (czgdp1807) [reply sympy/tensor/array/ndim_array.py]: It's not about convention, it's just about avoiding the things which aren't needed. I don't know if you follow some other principles.

Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: I changed the assignment of values as other blocs, so problem solved. 
But just out of curiousity, would you please explain to me why should elif -> if? I don't really see the point. As far as I know, a reasonable if-elif-else bloc ensures the logical consistency so that the code is more readable.

Comment (czgdp1807) [reply sympy/tensor/array/ndim_array.py]: No issues if you follow other principles. Everyone has their own coding style. It appeared non-intuitive to me to have an if just after an elif.

File: sympy/tensor/array/sparse_ndim_array.py [hunk 9]
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
...
-    def _eval_derivative_n_times(self, s, n):
-        from sympy import Integer, Derivative
-        if isinstance(n, (int, Integer)):
+    def _eval_derivative_array(self, arg):
+        from sympy.core.compatibility import Iterable
+        from sympy.matrices.common import MatrixCommon
+        from sympy import Derivative
+        from sympy import Dict
+        if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
+            return type(self)([self.diff(x) for x in arg])
```
Comment (kangzhiq) [inline sympy/tensor/array/sparse_ndim_array.py]: This case still cast sparse array to dense array. I have tried `derive_by_array` but `derive_by_array` itself has the same problem. Maybe an algorithm to calculate the new index after concatenating some sparse arrays, so that we can initiate a sparse array with the passed dictionary. That is to say:

[self.diff(x) for x in arg] -> [a, b, c] where a, b, c are sparse arrays. 

The question is how to calculate the index of non-zero value of a, b, c in a new dictionary, say D, so that D can represent [a, b, c].

File: sympy/tensor/array/sparse_ndim_array.py [hunk 10]
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
...
+        if not isinstance(other, NDimArray):
+            return False
+
+        if self.shape != other.shape:
+            return False
+
+        if isinstance(other, SparseNDimArray):
+            from sympy import Dict
+            dict_self = Dict(self._sparse_array)
+            dict_other = Dict(other._sparse_array)
```
Comment (kangzhiq) [inline sympy/tensor/array/sparse_ndim_array.py]: The dictionaries are cast to `Dict` because SymPy 's `Dict` and Python's built-in `dict` are not fully compatible. By saying that I mean that 
```python
>>> from sympy import Dict
>>> a = Dict({3:5, 2:7})
>>> b = {3:5, 2:7}
>>> a == b
False
```
They are not equal although they have exactly the same value.

File: sympy/tensor/array/sparse_ndim_array.py [hunk 11]
```diff
@@ -111,19 +111,40 @@ def reshape(self, *newshape):
...
         return type(self)(*(newshape + (self._sparse_array,)))
 
-    def _eval_derivative_n_times(self, s, n):
-        from sympy import Integer, Derivative
-        if isinstance(n, (int, Integer)):
+    def _eval_derivative_array(self, arg):
+        from sympy.core.compatibility import Iterable
+        from sympy.matrices.common import MatrixCommon
+        from sympy import Derivative
+        from sympy import Dict
```
Comment (kangzhiq) [inline sympy/tensor/array/sparse_ndim_array.py]: Imported API may need regrouped. Waiting for some advices. :-)

File: sympy/tensor/array/sparse_ndim_array.py [hunk 12]
```diff
@@ -109,8 +109,42 @@ def reshape(self, *newshape):
...
+            s_a = self._sparse_array
+            s_a2 = {}
+            for key in s_a:
+                value = _sympify(s_a[key])
+                deri = Derivative(value, arg, **{'evaluate':True})
+                if deri != 0:
+                    s_a2[key] = deri
+            return type(self)(Dict(s_a2), self._shape)
+
+    def __eq__(self, other):
```
Comment (kangzhiq) [inline sympy/tensor/array/sparse_ndim_array.py]: Since the evaluation of derivative triggers `__eq__`, this issue is handled together in this PR. I don't know if it is appropriate to do so. Waiting for some advices. :-)

Comment (Upabjojr) [reply sympy/tensor/array/sparse_ndim_array.py]: What about temporarily ignore equality checks?

File: sympy/tensor/array/sparse_ndim_array.py (hunk 10 above)
Comment (Upabjojr) [reply sympy/tensor/array/sparse_ndim_array.py]: No casting please. There are better ways to compare.

File: sympy/tensor/array/ndim_array.py [hunk 13]
```diff
@@ -129,34 +129,40 @@ def f(pointer):
...
-            shape = iterable.shape
-
-        # Construct N-dim array from another N-dim array:
-        elif shape is None and isinstance(iterable, NDimArray):
-            shape = iterable.shape
+        from sympy.tensor.array import SparseNDimArray
+        from sympy import Dict
+
+        if shape is None:
+            if iterable is None:
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: is this change necessary?

**Comment by Upabjojr** [comment]
Tests are needed.

**Comment by Upabjojr** [comment]
Generally, you should not create new methods for equality checks. There is already one I believe, you should probably work on that one.

File: sympy/tensor/array/sparse_ndim_array.py (hunk 12 above)
Comment (kangzhiq) [reply sympy/tensor/array/sparse_ndim_array.py]: @Upabjojr Do you mean that we keep equality checks casting sparse array to dense array and handle the case of `__eq__` in another PR?

File: sympy/tensor/array/ndim_array.py (hunk 13 above)
Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: This change is to refactor the code so that ```shape is None``` will not be repeated several times

File: sympy/tensor/array/sparse_ndim_array.py (hunk 12 above)
Comment (Upabjojr) [reply sympy/tensor/array/sparse_ndim_array.py]: There is already a method handling __eq__ in the superclass. I think the code should remain there.

File: sympy/tensor/array/ndim_array.py (hunk 13 above)
Comment (Upabjojr) [reply sympy/tensor/array/ndim_array.py]: OK, open a new PR next time. A separate PR on this issue would already be mergeable.

File: sympy/tensor/array/arrayop.py [hunk 14]
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
...
             return diff(expr, dx)
 
+def _sparse_array_derive_by_array(expr, dx):
+    # Perfom a derivative on a sparse array by each term of dx. Regroup the returned
+    # diff sprase array by each term to a dictionary.
+
+    sparse_array = {}
+    loop_size = len(expr)
+    for i, x in enumerate(dx):
+        sparse_diff = derive_by_array(expr, x)._sparse_array
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: why is this calling derive_by_array?

File: sympy/tensor/array/arrayop.py [hunk 15]
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
 
     """
     from sympy.matrices import MatrixBase
+    from sympy.tensor import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
-
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: why was this line remove?

File: sympy/tensor/array/arrayop.py [hunk 16]
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
...
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
+                new_array = _sparse_array_derive_by_array(expr, dx)
+            else:
+                new_array = [[y.diff(x) for y in expr] for x in dx]
             return type(expr)(new_array, dx.shape + expr.shape)
         else:
-            return expr.diff(dx)
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: This line was correct. IIRC, this depends on the applyfunc method, so just fix that method.

File: sympy/tensor/array/arrayop.py [hunk 17]
```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
...
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
+        if isinstance(expr, NDimArray):
+            expr = expr.as_immutable()
+        else:
+            expr = ImmutableDenseNDimArray(expr)
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: if both expr and dx are sparse, just use a dict comprehension (in a single line).

File: sympy/tensor/array/ndim_array.py [hunk 18]
```diff
@@ -129,34 +129,40 @@ def f(pointer):
...
+            elif isinstance(iterable, NDimArray):
+                shape = iterable.shape
+
+            else:
+                shape = ()
+                iterable = (iterable,)
 
         # Construct NDimArray(iterable, shape)
         elif shape is not None:
             pass
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: if you remove "else" under this line, this code becomes useless.

File: sympy/tensor/array/ndim_array.py [hunk 19]
```diff
@@ -420,9 +426,19 @@ def __eq__(self, other):
...
+
+        if not self.shape == other.shape:
+            return False
+
+        if isinstance(self, SparseNDimArray) and isinstance(other, SparseNDimArray):
+            return self._sparse_array.keys() == other._sparse_array.keys() and \
+                   all(v == other._sparse_array[k] for (k, v) in self._sparse_array.items()
+                       if k in other._sparse_array)
+
+        return list(self) == list(other)
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: why not just use something similar to this? dict(...) == dict(... other ...) ?

File: sympy/tensor/array/sparse_ndim_array.py [hunk 20]
```diff
@@ -109,8 +109,7 @@ def reshape(self, *newshape):
         if new_total_size != self._loop_size:
             raise ValueError("Invalid reshape parameters " + newshape)
 
-        return type(self)(*(newshape + (self._array,)))
-
+        return type(self)(*(newshape + (self._sparse_array,)))
```
Comment (Upabjojr) [inline sympy/tensor/array/sparse_ndim_array.py]: is this a bugfix?

File: sympy/tensor/array/tests/test_arrayop.py [hunk 21]
```diff
@@ -84,6 +85,13 @@ def test_derivative_by_array():
     assert diff(Array([[x, y], [z, t]]), Array([[x, y], [z, t]])) == Array([[[[1, 0], [0, 0]], [[0, 1], [0, 0]]],
                                                                          [[[0, 0], [1, 0]], [[0, 0], [0, 1]]]])
 
+    # test for large scale sparse array
+    b = MutableSparseNDimArray.zeros(10000, 20000)
+    b[0, 0] = i
+    b[0, 1] = j
+    assert derive_by_array(b, i)._sparse_array == Dict({0: 1})
```
Comment (Upabjojr) [inline sympy/tensor/array/tests/test_arrayop.py]: what about comparing it to an array?

File: sympy/tensor/array/arrayop.py (hunk 15 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: I have seen other examples in NDimArray that there are not new lines between the `import `and the executive code, which made me feel like it is ok to remove it.

File: sympy/tensor/array/arrayop.py (hunk 17 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: The line 213 uses dict comprehension when expr is sparse but dx is not. 
I think for the case where `dx` is an array(sparse or dense), it is necessary to calculate the new index for the derivative by each term in `dx` , which is the purpose of `_sparse_array_derive_by_array`.

File: sympy/tensor/array/ndim_array.py (hunk 19 above)
Comment (kangzhiq) [reply sympy/tensor/array/ndim_array.py]: I thought I am not supposed to cast one type to the other so I came up with this code. But if we can cast `Dict` to `dict`, that would indeed be much simpler. :-)
By the way, the problem of comparing `Dict` and `dict` will dispear if the test about `mutableArray.diff() == MutableArray` is changed to `mutableArray.diff() == ImmutableArray()`.

File: sympy/tensor/array/sparse_ndim_array.py (hunk 20 above)
Comment (kangzhiq) [reply sympy/tensor/array/sparse_ndim_array.py]: Yes, but apprently the problem isn't resolved. I will discard this change and fix this issue in a seperated PR.

File: sympy/tensor/array/arrayop.py (hunk 16 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: I think it is a little bit compliacted to modify `applyfunc.` My initial idea is to apply the function only to the dictionary of sprase array, in this case, `diff`, it makes sense. But when it comes to some operations which change 0 to other value, e.g. `lambda x: x+1` , every value should be added by one, not only the ones in the dictionary.
So I only modified the function which triggers applyfunc.

Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: I am trying with a test like this to make sure 0 is not changed:
```python
if f(0) == 0 and isinstance(self, SparseNDimArray):
        return type(self)({k: f(v) for k, v in my_dictionary.items()},self.shape)
```
But f(0) raised several errors. Do you have any idea to find a bypath for this test?

File: sympy/tensor/array/arrayop.py (hunk 14 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: This is because `.diff()` was slow when it comes to large scale sparse array, so I chose not to use it. Besides, `derive_by_array `, as said in the description of the function, supports the derivate by scalars, so I used it. If `.diff()` or `applyfunc` is fixed, then we can simply call `expr.diff(x)`.

File: sympy/tensor/array/ndim_array.py (hunk 19 above)
Comment (Upabjojr) [reply sympy/tensor/array/ndim_array.py]: Just cast to dict.

File: sympy/tensor/array/arrayop.py (hunk 17 above)
Comment (Upabjojr) [reply sympy/tensor/array/arrayop.py]: Yes, so why not just an array comprehension?

Comment (Upabjojr) [reply sympy/tensor/array/arrayop.py]: `{k1+k2: v2.diff(v1) for k1, v1 in dx... dict ...items() for k2, v2 in expr... dict ...items()}`

File: sympy/tensor/array/ndim_array.py [hunk 22]
```diff
@@ -251,10 +253,15 @@ def _eval_derivative(self, arg):
     def _eval_derivative_array(self, arg):
         from sympy import derive_by_array
         from sympy import Tuple
+        from sympy import SparseNDimArray
         from sympy.matrices.common import MatrixCommon
+
         if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
             return derive_by_array(self, arg)
         else:
+            if isinstance(self, SparseNDimArray):
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: This method was not supposed to be modified, applyfunc should be able to handle sparse arrays efficiently.

File: sympy/tensor/array/arrayop.py (hunk 17 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: OK, I got your point! Thank you!

File: sympy/tensor/array/arrayop.py [hunk 23]
```diff
@@ -197,9 +198,20 @@ def derive_by_array(expr, dx):
...
+                lp = len(expr)
+                new_array = {k + i*lp: expr.diff(x)._sparse_array[k]
+                             for k in expr._sparse_array
+                             for i, x in enumerate(dx)
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: why not iterate over the sparse array? You are flattening all members and then filtering with an if on the next line.

File: sympy/tensor/array/arrayop.py [hunk 24]
```diff
@@ -209,7 +221,6 @@ def derive_by_array(expr, dx):
         else:
             return diff(expr, dx)
 
-
```
Comment (Upabjojr) [inline sympy/tensor/array/arrayop.py]: leave two blank lines separation in the zeroth indentation.

File: sympy/tensor/array/sparse_ndim_array.py [hunk 25]
```diff
@@ -111,7 +111,6 @@ def reshape(self, *newshape):
 
         return type(self)(*(newshape + (self._array,)))
 
-
```
Comment (Upabjojr) [inline sympy/tensor/array/sparse_ndim_array.py]: two blank lines in zeroth indentation

File: sympy/tensor/array/ndim_array.py [hunk 26]
```diff
@@ -270,6 +274,11 @@ def applyfunc(self, f):
         >>> m.applyfunc(lambda i: 2*i)
         [[0, 2], [4, 6]]
         """
+        from sympy.tensor.array import SparseNDimArray
+
+        if isinstance(self, SparseNDimArray) and f(sympify(0)) == 0:
```
Comment (Upabjojr) [inline sympy/tensor/array/ndim_array.py]: S.Zero is the SymPy zero.

File: sympy/tensor/array/tests/test_mutable_ndim_array.py [hunk 27]
```diff
@@ -131,6 +131,14 @@ def test_sparse():
...
     assert sparse_array[0, 0] == 0
 
+    # test for large scale sparse array
+    a = MutableSparseNDimArray.zeros(100000, 200000)
+    b = MutableSparseNDimArray.zeros(100000, 200000)
+    assert a == b
+    a[1, 1] = 1
+    b[1, 1] = 2
+    assert a != b
+
```
Comment (Upabjojr) [inline sympy/tensor/array/tests/test_mutable_ndim_array.py]: test the behaviour of .applyfunc(fun) where fun(0) != 0

File: sympy/tensor/array/arrayop.py (hunk 23 above)
Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: This is because dx is an arraytype, not exactly a sparse array. Besides, I was allowing flatterning of `dx` because `dx` is supposed to be a sequence of non-zero values because 
```
>>> S.Zero._diff_wrt
False
```
So I think `dx` is not supposed to be a sparse array. Am I misunderstanding something here? Please let me know if so, thank you!!

Comment (kangzhiq) [reply sympy/tensor/array/arrayop.py]: I think I got your point! Working on it.
